Scroll and flash text indicator for fragment when loading scroll to text fragment URL

megangardner opened the pull request:
#### c6891f9bece8cee95626f79d9199dad21ca65de9
<pre>
Scroll and flash text indicator for fragment when loading scroll to text fragment URL
<a href="https://bugs.webkit.org/show_bug.cgi?id=243618">https://bugs.webkit.org/show_bug.cgi?id=243618</a>
rdar://83902653 (Separate scroll and flash for noting highlights on load)

Reviewed by Tim Horton.

Add code to properly scroll to the fragment that we have found as well as add
a delay for flashing the text indicator when we arrive at that location.
Previously, especially on Mac, if we created the text indicator immediately, it
would not show up unless we were on the first &apos;page&apos; of the website, and a scroll
did not need to happen. Otherwise, it was create at the offset of the range before
the scroll, resulting in an flash in a different location to the range. Even if it
was in the correct spot, it was very early, not leading to a very polished visual for
the user. Add a small delay, and the visual improves significantly.
Disable the indicator for ref-tests, as it can cause issues with testing.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::FrameView):
(WebCore::FrameView::reset):
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::maintainScrollPositionAtAnchor):
(WebCore::FrameView::setScrollPosition):
(WebCore::FrameView::textFragmentIndicatorTimerFired):
(WebCore::FrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::FrameView::scrollToAnchor):
(WebCore::FrameView::setWasScrolledByUser):
* Source/WebCore/page/FrameView.h:

Canonical link: <a href="https://commits.webkit.org/253327@main">https://commits.webkit.org/253327@main</a>
</pre>

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2320,6 +2324,7 @@ void FrameView::maintainScrollPositionAtAnchor(ContainerNode* anchorNode)
         return;
 
     cancelScheduledScrollToFocusedElement();
```
I wonder if we should factor all of these out into "cancel scheduled scrolls" function so we only have to add them in one place? But maybe this is OK.

hortont424 replied:
... but now that I scroll down, there are so many... so maybe do!

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2451,6 +2457,29 @@ void FrameView::scrollToFocusedElementInternal()
+    if (!m_textFragmentIndicatorTimerRange)
+        return;
+    
+    auto textIndicator = TextIndicator::createWithRange(m_textFragmentIndicatorTimerRange.value(), { TextIndicatorOption::DoNotClipToVisibleRect }, WebCore::TextIndicatorPresentationTransition::Bounce);
```
Do you need to do anything to make sure the range is still valid? Or maybe TextIndicator does it for you?

hortont424 commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2451,6 +2457,29 @@ void FrameView::scrollToFocusedElementInternal()
+    if (textIndicator)
+        document.page()->chrome().client().setTextIndicator(textIndicator->data());
+    
+    m_textFragmentIndicatorTimerRange.reset();
```
I don't think this should have "timer" in the name. Maybe "pendingTextFragmentIndicatorRange"?

rniwa commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2241,8 +2244,11 @@ bool FrameView::scrollToFragment(const URL& url)
+                TemporarySelectionChange selectionChange(document, { range }, { TemporarySelectionOption::DelegateMainFrameScroll, TemporarySelectionOption::RevealSelectionBounds, TemporarySelectionOption::UserTriggered });
+                m_pendingTextFragmentIndicatorRange = range;
+                m_pendingTextFragmentIndicatorText = plainText(range);
+                m_delayedTextFragmentIndicatorTimer.startOneShot(100_ms);
```
100ms seems really slow. Humans can perceive that. Do we really need to wait that long?

rniwa replied:
For example, HTML5 event loop defines when we fire scroll events:
https://html.spec.whatwg.org/multipage/webappapis.html#update-the-rendering

Perhaps that is more appropriate time to wait for?

hortont424 replied:
This is a human-factors timer (load+scroll -> take a beat -> flash), not here for a technical reason, so it's good that it's a perceptible time, IMO :)

megangardner replied:
I tried 100ms and 250ms, I felt that 100 was better, but HI is going to be the real decision maker here, I didn't want to spend too much time on a number that was very likely to be overridden.
And yes, it should be human-perceivable, as Tim said. We _want_ to make sure that it can be seen.

rniwa commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2451,6 +2463,34 @@ void FrameView::scrollToFocusedElementInternal()
+    if (textIndicator)
+        document.page()->chrome().client().setTextIndicator(textIndicator->data());
+    
+    m_pendingTextFragmentIndicatorRange.reset();
```
Why don't we just call cancel to share code?

rniwa commented on Source/WebCore/page/FrameView.cpp:
```diff
@@ -2401,6 +2407,12 @@ void FrameView::scheduleScrollToFocusedElement(SelectionRevealMode selectionReve
 
+void FrameView::cancelScheduledScrolls()
+{
+    cancelScheduledScrollToFocusedElement();
```
What happens when this is canceled but the other isn't? It seems a bit fragile that this relies on every call site of this function to be also calling the other function.

If we should always be calling the new function instead, perhaps we should make this private.

megangardner replied:
The whole point of this function was to tie these together, as Tim suggested. I'm unsure if there is a reason to cancel one and not the other, so I did not go so far as to make it private.

hortont424 replied:
(but I think you can make the two functions it *calls* private)

megangardner committed:
Scroll and flash text indicator for fragment when loading scroll to text fragment URL
https://bugs.webkit.org/show_bug.cgi?id=243618
rdar://83902653 (Separate scroll and flash for noting highlights on load)

Reviewed by Tim Horton.

Add code to properly scroll to the fragment that we have found as well as add
a delay for flashing the text indicator when we arrive at that location.
Previously, especially on Mac, if we created the text indicator immediately, it
would not show up unless we were on the first 'page' of the website, and a scroll
did not need to happen. Otherwise, it was create at the offset of the range before
the scroll, resulting in an flash in a different location to the range. Even if it
was in the correct spot, it was very early, not leading to a very polished visual for
the user. Add a small delay, and the visual improves significantly.
Disable the indicator for ref-tests, as it can cause issues with testing.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::FrameView):
(WebCore::FrameView::reset):
(WebCore::FrameView::scrollToFragment):
(WebCore::FrameView::maintainScrollPositionAtAnchor):
(WebCore::FrameView::setScrollPosition):
(WebCore::FrameView::textFragmentIndicatorTimerFired):
(WebCore::FrameView::cancelScheduledTextFragmentIndicatorTimer):
(WebCore::FrameView::scrollToAnchor):
(WebCore::FrameView::setWasScrolledByUser):
* Source/WebCore/page/FrameView.h:

Canonical link: https://commits.webkit.org/253327@main
